tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go

```diff
@@ -160,6 +191,17 @@ func TestAuthJwksAsyncFetch(t *testing.T) {
 			afterProviderCnt: &expectedRequestCount{provider, 5},
 			wantResp:         `{"aud":["admin.cloud.goog","bookstore_test_client.cloud.goog"],"exp":4698318999,"iat":1544718999,"iss":"api-proxy-testing@cloud.goog","sub":"api-proxy-testing@cloud.goog"}`,
 		},
+		{
+			desc:                "Success, the jwks provider succeeds only once every 5 requests so at least 5 request to the jwks provider will be made",
+			path:                "/auth/info/auth0",
+			apiKey:              "api-key",
+			method:              "GET",
+			jwksFetchNumRetries: 5,
+			token:               testdata.FakeCloudTokenMultiAudiences,
+			initProviderCnt:     &expectedRequestCount{provider, 1},
+			afterProviderCnt:    &expectedRequestCount{provider, 5},
```

Comment:
Hmm,  for async_fetch,  all 5 retry should happen before listener is activated.  or before Envoy is ready,  it it should be all in "initProviderCnt".  Could you help to debug it to see why that is not the case

You can run the test by
```
go test -v ./tests/integration_test/auth_pkey_cache_test/auth_pkey_cache_test.go --debug_components=envoy --logtostderr &> /tmp/mmm.log
```

Reply:
the test done on initProviderCount was "I received at least one request" not "I received exactly one request".
If I log the exact number of request obtained it is indeed 5. And I should test for that.

with a provider in this test 
```
 {
  key: "google_jwt"
  value {
    issuer: "api-proxy-testing@cloud.goog"
    audiences: "https://echo-api.endpoints.cloudesf-testing.cloud.goog"
    remote_jwks {
      http_uri {
        uri: "http://127.0.0.1:20113"
        cluster: "jwt-provider-cluster-127.0.0.1:20113"
        timeout {
          seconds: 30
        }
      }
      cache_duration {
        seconds: 300
      }
      async_fetch {
      }
      retry_policy {
        retry_back_off {
          base_interval {
            nanos: 200000000
          }
          max_interval {
            seconds: 32
          }
        }
        num_retries {
          value: 5
        }
      }
    }
    forward: true
    from_headers {
      name: "Authorization"
      value_prefix: "Bearer "
    }
    from_headers {
      name: "X-Goog-Iap-Jwt-Assertion"
    }
    from_params: "access_token"
    forward_payload_header: "X-Endpoint-API-UserInfo"
    payload_in_metadata: "jwt_payloads"
  }
}
```

It wasn't clear to me that the envoy proxy ingress listener would attempt all retries to its provider JwksFetcher during the set up... I thought I could only guarantee one call at minimum.
Then upon success of a JWT-authorized request, I could guarantee all five retries had occurred, the last one returning a proper JWKS, and the JWT being validated...

I'll modify the test's initProviderCnt accordingly.

Reply:
I see. If verifyProviderCnt() is checking minimum request count,  change initProviderCnt to 5 is good enough.  Thanks